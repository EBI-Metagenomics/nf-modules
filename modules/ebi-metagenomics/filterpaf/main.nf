process FILTERPAF {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/gawk:5.3.0' :
        'biocontainers/gawk:5.3.0' }"

    input:
    tuple val(meta), path(paf_file)

    output:
    tuple val(meta), path("${prefix}.txt"), emit: mapped_contigs_txt
    path "versions.yml"                    , emit: versions

    script:
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Filter PAF by query coverage and MAPQ
    awk '
        {
            query_len = \$2;
            query_start = \$3;
            query_end = \$4;
            target = \$6;
            matching_bases = \$10;
            total_bases = \$11;

            if (target == "*") {
                next;
            }

            aligned_len = query_end - query_start;
            query_coverage = aligned_len / query_len;
            pid = matching_bases / total_bases;

            if (query_coverage >= ${params.min_qcov} && pid >= ${params.min_pid}) {
                print \$1;
            }
        }
    ' ${paf_file} > ${prefix}.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        awk: \$(awk --version 2>&1 | grep -oE '[0-9]{8}')
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        awk: \$(awk --version 2>&1 | grep -oE '[0-9]{8}')
    END_VERSIONS
    """
}
