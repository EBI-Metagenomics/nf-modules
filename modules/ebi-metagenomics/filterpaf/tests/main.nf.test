
nextflow_process {

    name "Test Process FILTERPAF"
    script "../main.nf"
    process "FILTERPAF"

    tag "modules"
    tag "modules_ebimetagenomics"
    tag "filterpaf"

    test("filterpaf") {

        when {
            params {
                min_qcov = 0.3
                min_pid  = 0.4
            }
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file("${moduleDir}/tests/data/test.paf", checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("filterpaf - stub") {

        options "-stub"

        when {
            params {
                min_qcov = 0.3
                min_pid  = 0.4
            }
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file("${moduleDir}/tests/data/test.paf", checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
