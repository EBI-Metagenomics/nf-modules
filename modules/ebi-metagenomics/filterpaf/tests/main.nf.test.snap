{
    "filterpaf - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,394280fc078876591e1dab58866e90de"
                ],
                "mapped_contigs_txt": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,394280fc078876591e1dab58866e90de"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.1"
        },
        "timestamp": "2025-05-22T14:18:58.976714528"
    },
    "filterpaf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,51e09c5bb215ecd7e9bf0ad82316f13b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,394280fc078876591e1dab58866e90de"
                ],
                "mapped_contigs_txt": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,51e09c5bb215ecd7e9bf0ad82316f13b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,394280fc078876591e1dab58866e90de"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-28T09:02:54.103907"
    }
}